arangod/Agency/Supervision.cpp

```diff
@@ -1121,32 +1132,62 @@ void Supervision::workJobs() {
   auto todos = _snapshot.hasAsChildren(toDoPrefix).first;
   auto it = todos.begin();
   static std::string const FAILED = "failed";
-    
+
+  size_t const maximalJobsPerRound = 1000;
+  bool selectRandom = todos.size() > maximalJobsPerRound;
+
+  LOG_TOPIC(TRACE, Logger::SUPERVISION) << "Begin ToDos of type Failed*";
   while (it != todos.end()) {
+    if (selectRandom && rand() % todos.size() > maximalJobsPerRound) {
```

Comment:
Actually, this is not true. We are working on a snapshot of the agency which is not changed as the loop progresses. Agency transactions are performed which might reduce the number of ToDos, but this will not be reflected in the snapshot. So I think this is mathematically correct.
In any case, it is just a rule of thumb to do only 1000 out of all. We can approximately perform 200 jobs per second, so the 1000 achieves that after at most 5 seconds we do another round and, for example, check if a server is good again. So I think this is somehow sensible. The 1000 is of course just hand picked. One could have chosen 500 or 1500. Over time, we might be able to optimize this.